californium-core/src/main/java/org/eclipse/californium/core/network/Exchange.java

```diff
@@ -583,10 +586,13 @@ public void setCustomExecutor() {
 	 * @param ctx the correlation information
 	 */
 	public void setCorrelationContext(final CorrelationContext ctx) {
-		correlationContext = ctx;
-		ExchangeObserver obs = this.observer;
-		if (obs != null) {
-			obs.contextEstablished(this);
+		if (correlationContext.compareAndSet(null, ctx)) {
+			ExchangeObserver obs = this.observer;
+			if (obs != null) {
+				obs.contextEstablished(this);
+			}
+		} else {
+			correlationContext.set(ctx);
```

Comment:
I think you put too many hope on this discussion ;).
But I'm ok to let this point beside for now.